sklearn/utils/tests/test_stats.py

```diff
@@ -96,3 +96,71 @@ def test_weighted_percentile_2d():
         _weighted_percentile(x_2d[:, i], w_2d[:, i]) for i in range(x_2d.shape[1])
     ]
     assert_allclose(w_median, p_axis_0)
+
+
+def test_weighted_percentile_nan_filtered():
+    """Test that calling _weighted_percentile on an array with nan values returns
+    the same results as calling _weighted_percentile on a filtered version of the data.
+    We test both with sample_weight of the same shape as the data and for
+    one-dimensional sample_weight."""
+
+    rng = np.random.RandomState(42)
+    array = rng.rand(10, 100)
+
+    nan_array = array.copy()
+    nan_array[rng.rand(*nan_array.shape) < 0.5] = np.nan
+    nan_mask = np.isnan(nan_array)
+
+    sample_weights = [rng.randint(1, 6, size=(10, 100)), rng.randint(1, 6, size=(10,))]
+
+    for sample_weight in sample_weights:
+        # Find the weighted percentile on the array with nans:
+        values_nan = _weighted_percentile(nan_array, sample_weight, 30)
+
+        # Find the weighted percentile on the filtered array:
+        filtered_values = [
+            array[~nan_mask[:, col], col] for col in range(array.shape[1])
+        ]
+        if sample_weight.ndim == 1:
+            sample_weight = np.repeat(sample_weight, array.shape[1]).reshape(
+                array.shape[0], array.shape[1]
+            )
+        filtered_weights = [
+            sample_weight[~nan_mask[:, col], col] for col in range(array.shape[1])
+        ]
+
+        values_filtered = np.array(
+            [
+                _weighted_percentile(filtered_values[col], filtered_weights[col], 30)
+                for col in range(array.shape[1])
+            ]
+        )
+
+        assert_array_equal(values_filtered, values_nan)
```

Comment:
I would do the following variable renaming to help make this code easier to understand:

- `values_nan` renamed to `results`
- `values_filtered` renamed to `expected_results`
- `filtered_values` renamed to `filtered_arrays`

Reply:
I really like that. Thank you.